bin/identify_dormant_github_users.py

```diff
@@ -32,48 +41,156 @@
+                if username not in bot_list and not is_collaborator:
+                    usernames.append(username)
+
+    except FileNotFoundError as e:
+        logging.error("Error reading from file %s: %s", CSV_FILE_NAME, e)
+
+    return usernames
+
+
+def download_github_dormant_users_csv_from_s3():
+    s3 = boto3.client("s3")
+    try:
+        s3.download_file(BUCKET_NAME, CSV_FILE_NAME, CSV_FILE_NAME)
+        logger.info("File %s downloaded successfully.", CSV_FILE_NAME)
+    except NoCredentialsError:
+        logger.error(
+            "Credentials not available, please check your AWS credentials.")
+    except FileNotFoundError as e:
+        logger.error("Error downloading file: %s", e)
+
+    if not os.path.isfile(CSV_FILE_NAME):
+        raise FileNotFoundError(
+            f"File {CSV_FILE_NAME} not found in the current directory."
+        )
+
 
-    cloudtrail_service = CloudtrailService()
-    active_users = cloudtrail_service.get_active_users_for_dormant_users_process()
+def get_dormant_users_from_github_csv(
+    moj_github_org: GithubService, ap_github_org: GithubService
+) -> list[DormantUser]:
+    """
+    This function depends on a preliminary manual process: a GitHub Enterprise user
+    must first download the 'dormant.csv' file from the GitHub Enterprise dashboard and
+    then upload it to an S3 bucket. Once this setup is complete, this function will
+    download the 'dormant.csv' file from the S3 bucket and extract a list of
+    usernames from the file.
 
-    return [user for user in all_users if user not in list(set(active_users).union(bot_list))]
+    This process ensures that the most current data regarding dormant users is used.
+    """
+    download_github_dormant_users_csv_from_s3()
+    users = get_usernames_from_csv_ignoring_bots_and_collaborators(
+        ALLOWED_BOT_USERS)
 
+    dormant_users = [
+        DormantUser(
+            user,
+            moj_github_org.get_user_org_email_address(user)
+            or ap_github_org.get_user_org_email_address(user),
+        )
+        for user in users
+    ]
 
-def get_dormant_users_from_data_lake(github_service: GithubService) -> list:
-    list_of_inactive_github_users = get_inactive_users_from_data_lake_ignoring_bots_and_collaborators(github_service, ALLOWED_BOT_USERS)
+    return dormant_users
 
-    list_of_dormant_users = [DormantGitHubUser(github_service, user) for user in list_of_inactive_github_users]
 
-    return list_of_dormant_users
+def get_active_users_from_auth0_log_group() -> list:
+    """Operations Engineering, Cloud, Mod and Analytical Platforms all have thir own
+    Auth0 tenants. All logs are collected in a single Cloudwatch log group.
+    This function will parse the logs and return a list of active users using
+    an insights query.
+    """
+    cloudwatch_log_group = "/aws/events/LogsFromOperationsEngineeringAuth0"
 
+    cloudwatch_service = CloudWatchService(cloudwatch_log_group)
```

Comment:
Similar here, would prefer the service to be created first at the top level and injected into the relevant processes 💉